* Alignment
  - All the bam files are generated by snaptools, which is an old
    python package for SnapATAC R package.
  - In snaptools, bwa software is used for the algnment.
  - See details under src/main/pipeline/alignment.Snakefile
  - The config.json and cluster.json are under src/main/resource
    
* Quality Control and Doublet Removal
  - After generating bam files, we then use SnapATAC2 for our
    analysis.
  - See details under src/main/pipeline/snapatac2.qc.Snakefile.
  - The snapatac2.qc.config.yaml is under src/main/resource.
** Quality control
   - At bulk level, we check the quality of the sequencing based on
     the sequencing depth, number of unique fragments, and so on. The
     most important thing is to check if the size of fragments in each
     sample were enriched in 100 bp and 200 bp (see [[https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1929-3][ref]]).
   - At single-cell level, we follow the
     [[https://www.encodeproject.org/atac-seq/][ENCODE pipeline]], we use two criteria to filter cells:
     1. Number of unique fragment >= 1,000
     2. Transcription start site enrichment (TSSe) >= 10
** Doublet removal
   - We use Scrublet for this. And the imputed gene expression from
     snATAC-seq data is similar like what Seurat did.
